src/asr/asr_mc_chainer.py

```diff
@@ -0,0 +1,516 @@
+
+        # Compute the loss at this time step and accumulate it
+        loss = optimizer.target(x)
+        optimizer.target.cleargrads()  # Clear the parameter gradients
+        loss.backward()  # Backprop
+        loss.unchain_backward()  # Truncate the graph
+        # compute the gradient norm to check if it is normal or not
+        grad_norm = np.sqrt(self._sum_sqnorm(
+            [p.grad for p in optimizer.target.params(False)]))
+        logging.info('grad norm={}'.format(grad_norm))
+        if math.isnan(grad_norm):
+            logging.warning('grad norm is nan. Do not update model.')
+        else:
+            optimizer.update()
+        delete_feat(x)
+
+    # copied from https://github.com/chainer/chainer/blob/master/chainer/optimizer.py
+    def _sum_sqnorm(self, arr):
+        sq_sum = collections.defaultdict(float)
+        for x in arr:
+            with cuda.get_device_from_array(x) as dev:
+                if x is not None:
+                    x = x.ravel()
+                    s = x.dot(x)
+                    sq_sum[int(dev)] += s
+        return sum([float(i) for i in six.itervalues(sq_sum)])
+
+
+def me2e_converter_train(batch, readers):
+    data = batch[0]
+    mode = data[1]['mode']
+
+    for data in batch:
+        # noisy mode
+        if mode == 'noisy':
+            bidim = readers['noisy'][data[0].encode('ascii', 'ignore')].shape[1] / 2
+            # separate real and imaginary part
+            feat_real = readers['noisy'][data[0].encode('ascii', 'ignore')][:, :bidim]
+            feat_imag = readers['noisy'][data[0].encode('ascii', 'ignore')][:, bidim:]
+        # enhancement mode
+        elif mode == 'enhan':
+            bidim = readers['enhan'][0][data[0].encode('ascii', 'ignore')].shape[1] / 2
+            # separate real and imaginary part
+            feat_real = [reader[data[0].encode('ascii', 'ignore')][:, :bidim] for reader in readers['enhan']]
+            feat_imag = [reader[data[0].encode('ascii', 'ignore')][:, bidim:] for reader in readers['enhan']]
+        else:
+            logging.error(
+                "Error: need to specify an appropriate training mode")
+            sys.exit()
+
+        feat = {}
+        feat['real'] = feat_real
+        feat['imag'] = feat_imag
+
+        data[1]['feat'] = feat
+
+    return batch
+
+
+def me2e_converter_recog(name, readers, mode):
+    # noisy mode
+    if mode == 'noisy':
+        bidim = readers[name].shape[1] / 2
+        # separate real and imaginary part
+        feat_real = readers[name][:, :bidim]
+        feat_imag = readers[name][:, bidim:]
+    # enhancement mode
+    elif mode == 'enhan':
+        bidim = readers[0][name].shape[1] / 2
+        # separate real and imaginary part
+        feat_real = [reader[name][:, :bidim] for reader in readers]
+        feat_imag = [reader[name][:, bidim:] for reader in readers]
+    else:
+        logging.error(
+            "Error: need to specify an appropriate decoding mode")
+        sys.exit()
```

Comment:
sys.exit() returns 0 that means success but it is error. so you need to change this line like
- `sys.exit(1)`
- `sys.exit("Error: need to specify an appropriate decoding mode")`

BTW I found that there are similar wrong usages in ESPnet...

Reply:
I see.
I will modify it.